File: backend/api/card_service.py
import requests

class YGOCardService:
    BASE_URL = "https://db.ygoprodeck.com/api/v7/cardinfo.php"

    @staticmethod
    def search_cards(query):
        """
        Busca cartas por nombre (fuzzy search) usando la API de YGOPRODeck.
        """
        if not query:
            return []
            
        params = {
            'fname': query,
            'num': 20, # Limitar a 20 resultados para rapidez
            'offset': 0
        }
        
        try:
            response = requests.get(YGOCardService.BASE_URL, params=params)
            response.raise_for_status()
            data = response.json()
            
            # Formatear la respuesta para el frontend
            results = []
            for card in data.get('data', []):
                results.append({
                    'id': card['id'],
                    'name': card['name'],
                    'type': card['type'],
                    'desc': card['desc'],
                    'atk': card.get('atk'),
                    'def': card.get('def'),
                    'level': card.get('level'),
                    'race': card['race'],
                    'attribute': card.get('attribute'),
                    'image_url': card['card_images'][0]['image_url'],
                    'image_url_small': card['card_images'][0]['image_url_small']
                })
            return results
        except requests.exceptions.RequestException as e:
            print(f"Error calling YGOPRODeck API: {e}")
            return []

    @staticmethod
    def get_cards_metadata(names):
        """
        Obtiene metadatos detallados para una lista de nombres de cartas.
        """
        if not names:
            return {}
            
        # Unir nombres con | para consulta múltiple
        # La API de YGOPRODeck acepta hasta un límite, así que lo ideal sería dividir si son muchos
        # Para un mazo de 40-60, una sola llamada suele bastar si la URL no es excesiva.
        
        chunk_size = 20 # Dividimos por seguridad de longitud de URL
        metadata_map = {}
        
        for i in range(0, len(names), chunk_size):
            chunk = names[i:i + chunk_size]
            params = {'name': '|'.join(chunk)}
            
            try:
                response = requests.get(YGOCardService.BASE_URL, params=params)
                if response.status_code == 200:
                    data = response.json()
                    for card in data.get('data', []):
                        metadata_map[card['name'].lower()] = {
                            'type': card['type'],
                            'image_url': card['card_images'][0]['image_url'],
                            'attribute': card.get('attribute'),
                            'race': card['race']
                        }
            except Exception as e:
                print(f"Error fetching metadata chunk: {e}")
                
        return metadata_map

    @staticmethod
    def find_best_match(name):
        """
        Busca la coincidencia más cercana para un nombre dado usando el endpoint 'fname'.
        """
        if not name:
            return None
            
        params = {
            'fname': name,
            'num': 1,
            'offset': 0
        }
        
        try:
            response = requests.get(YGOCardService.BASE_URL, params=params)
            if response.status_code == 200:
                data = response.json()
                if data.get('data'):
                    card = data['data'][0]
                    return {
                        'name': card['name'],
                        'type': card['type'],
                        'image_url': card['card_images'][0]['image_url'],
                        'attribute': card.get('attribute'),
                        'level': card.get('level'),
                        'atk': card.get('atk'),
                        'def': card.get('def')
                    }
        except Exception as e:
            print(f"Error in find_best_match: {e}")
            
        return None

File: api/_lib/card-service.ts
import axios from 'axios';

const BASE_URL = 'https://db.ygoprodeck.com/api/v7/cardinfo.php';

export interface YGOCard {
    id: number;
    name: string;
    type: string;
    desc: string;
    atk?: number;
    def?: number;
    level?: number;
    race: string;
    attribute?: string;
    image_url: string;
    image_url_small: string;
}

export interface CardMetadata {
    id: number;
    type: string;
    image_url: string;
    attribute?: string;
    race?: string;
    level?: number;
    atk?: number;
    def?: number;
}

export async function searchCards(query: string): Promise<YGOCard[]> {
    if (!query) return [];

    try {
        const response = await axios.get(BASE_URL, {
            params: { fname: query, num: 30, offset: 0 },
        });

        return (response.data.data || []).map((card: any) => ({
            id: card.id,
            name: card.name,
            type: card.type,
            desc: card.desc,
            atk: card.atk,
            def: card.def,
            level: card.level,
            race: card.race,
            attribute: card.attribute,
            image_url: card.card_images[0].image_url,
            image_url_small: card.card_images[0].image_url_small,
        }));
    } catch (error) {
        console.error('Error calling YGOPRODeck API:', error);
        return [];
    }
}

export async function getCardsMetadata(names: string[]): Promise<Record<string, CardMetadata>> {
    if (!names.length) return {};

    const chunkSize = 20;
    const metadataMap: Record<string, CardMetadata> = {};

    for (let i = 0; i < names.length; i += chunkSize) {
        const chunk = names.slice(i, i + chunkSize);
        try {
            const response = await axios.get(BASE_URL, {
                params: { name: chunk.join('|') },
            });

            if (response.status === 200) {
                for (const card of response.data.data || []) {
                    metadataMap[card.name.toLowerCase()] = {
                        id: card.id,
                        type: card.type,
                        image_url: card.card_images[0].image_url,
                        attribute: card.attribute,
                        race: card.race,
                    };
                }
            }
        } catch (error) {
            console.error('Error fetching metadata chunk:', error);
        }
    }

    return metadataMap;
}

export async function findBestMatch(name: string): Promise<CardMetadata & { name: string } | null> {
    if (!name) return null;

    try {
        const response = await axios.get(BASE_URL, {
            params: { fname: name, num: 1, offset: 0 },
        });

        if (response.status === 200 && response.data.data?.length) {
            const card = response.data.data[0];
            return {
                id: card.id,
                name: card.name,
                type: card.type,
                image_url: card.card_images[0].image_url,
                attribute: card.attribute,
                level: card.level,
                atk: card.atk,
                def: card.def,
            };
        }
    } catch (error) {
        console.error('Error in findBestMatch:', error);
    }

    return null;
}
